cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -524,21 +529,39 @@ public RunId runInternal(ProgramId programId, Map<String, String> userArgs, Map<
     userId = userId == null ? "" : userId;
 
     checkCapability(programDescriptor);
-    synchronized (this) {
-      if (maxConcurrentRuns > 0 && maxConcurrentRuns <= store.countActiveRuns(maxConcurrentRuns)) {
-        ConflictException e = new ConflictException(
-          String.format("Program %s cannot start because the maximum of %d concurrent runs is exceeded",
-                        programId, maxConcurrentRuns));
 
+    RunRecordCounter.Count count = runRecordCounter.addRequestAndGetCount(programId.run(runId));
+
+      if (maxConcurrentRunning > 0 && maxConcurrentRunning < count.getRunningCount()) {
+        TooManyRequestsException e = new TooManyRequestsException(
+          String.format("Program %s cannot start because the maximum of %d concurrent running runs is allowed",
+                        programId, maxConcurrentRunning));
+
+        runRecordCounter.removeRequest(programId.run(runId));
         programStateWriter.reject(programId.run(runId), programOptions, programDescriptor, userId, e);
         throw e;
       }
-    }
 
-    LOG.info("Attempt to run {} program {} as user {} with arguments {}", programId.getType(), programId.getProgram(),
-             authenticationContext.getPrincipal().getName(), userArgs);
+      if (maxConcurrentLaunching > 0 && maxConcurrentLaunching < count.getLaunchingCount()) {
+        TooManyRequestsException e = new TooManyRequestsException(
+          String.format("Program %s cannot start because the maximum of %d concurrent provisioning/starting runs" +
+                          " is allowed", programId, maxConcurrentLaunching));
+
+        runRecordCounter.removeRequest(programId.run(runId));
+        programStateWriter.reject(programId.run(runId), programOptions, programDescriptor, userId, e);
+        throw e;
+      }
+
+      LOG.info("Attempt to run {} program {} as user {} with arguments {}", programId.getType(), programId.getProgram(),
+               authenticationContext.getPrincipal().getName(), userArgs);
+
+    try {
+      provisionerNotifier.provisioning(programId.run(runId), programOptions, programDescriptor, userId);
+    } catch (Exception ex) {
+      runRecordCounter.removeRequest(programId.run(runId));
```

Comment:
It's better to do this in finally block. There are errors that rarely happen, but are not fatal (e.g. StackOverflowError) that will leave this counter in wrong state.

Reply:
Yep. So it's usually something like 
```
boolean done = false;
try {
 ...
 done = true;
finally {
 if (!done) {
  cleanup()
}
```

Reply:
Done!